File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/between/BetweenEnvelope.kt
package org.wyrdsekai.app.engine.between

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

/**
 * Wire format for all Between messages.
 * Signed with Ed25519 — signature covers src:dst:ts:payload.
 *
 * Mirrors server's BetweenEnvelope.java.
 */
@Serializable
data class BetweenEnvelope(
    val v: Int = 1,
    val src: String,
    val dst: String? = null,
    val ts: Long,
    val sig: String,
    val payload: JsonElement,
) {
    companion object {
        private val json = Json { ignoreUnknownKeys = true }

        /** Build the data that gets signed: "src:dst:ts:payload". */
        fun signingData(src: String, dst: String?, ts: Long, payload: JsonElement): ByteArray {
            val dstStr = dst ?: "*"
            val payloadStr = json.encodeToString(payload)
            return "$src:$dstStr:$ts:$payloadStr".encodeToByteArray()
        }

        /** Create and sign an envelope. */
        fun create(
            src: String,
            dst: String?,
            payload: JsonElement,
            identity: NodeIdentity,
        ): BetweenEnvelope {
            val ts = currentTimeMillis()
            val data = signingData(src, dst, ts, payload)
            val sig = identity.sign(data).encodeBase64()
            return BetweenEnvelope(v = 1, src = src, dst = dst, ts = ts, sig = sig, payload = payload)
        }

        /** Deserialize from JSON bytes. */
        fun fromBytes(data: ByteArray): BetweenEnvelope =
            json.decodeFromString(data.decodeToString())

        /** Current time in epoch milliseconds. */
        private fun currentTimeMillis(): Long =
            kotlin.time.Clock.System.now().toEpochMilliseconds()
    }

    /** Verify this envelope's signature against a peer's public key. */
    fun verify(peerPublicKey: ByteArray, crypto: CryptoProvider): Boolean {
        val data = signingData(src, dst, ts, payload)
        val sigBytes = sig.decodeBase64()
        return crypto.verify(peerPublicKey, data, sigBytes)
    }

    /** Serialize to JSON bytes. */
    fun toBytes(): ByteArray = json.encodeToString(this).encodeToByteArray()
}

// Base64 helpers (Kotlin Multiplatform)
@OptIn(kotlin.io.encoding.ExperimentalEncodingApi::class)
internal fun ByteArray.encodeBase64(): String =
    kotlin.io.encoding.Base64.encode(this)

@OptIn(kotlin.io.encoding.ExperimentalEncodingApi::class)
internal fun String.decodeBase64(): ByteArray =
    kotlin.io.encoding.Base64.decode(this)

File: clients/rn/src/engine/between/BetweenEnvelope.ts
/**
 * Wire format for all Between messages.
 * Signed with Ed25519 — signature covers src:dst:ts:payload.
 *
 * Mirrors server's BetweenEnvelope.java and KMP's BetweenEnvelope.kt.
 */

import type { NodeIdentity, CryptoProvider } from './NodeIdentity';

export interface BetweenEnvelope {
  v: number;
  src: string;
  dst: string | null;
  ts: number;
  sig: string;
  payload: unknown;
}

/** Build the data that gets signed: "src:dst:ts:payload". */
export function signingData(src: string, dst: string | null, ts: number, payload: unknown): Uint8Array {
  const dstStr = dst ?? '*';
  const payloadStr = JSON.stringify(payload);
  return new TextEncoder().encode(`${src}:${dstStr}:${ts}:${payloadStr}`);
}

/** Create and sign an envelope. */
export function createEnvelope(
  src: string,
  dst: string | null,
  payload: unknown,
  identity: NodeIdentity,
  crypto: CryptoProvider,
): BetweenEnvelope {
  const ts = Date.now();
  const data = signingData(src, dst, ts, payload);
  const sig = crypto.sign(identity.privateKey, data);
  return { v: 1, src, dst, ts, sig, payload };
}

/** Verify an envelope's signature against a peer's public key. */
export function verifyEnvelope(
  envelope: BetweenEnvelope,
  peerPublicKey: string,
  crypto: CryptoProvider,
): boolean {
  const data = signingData(envelope.src, envelope.dst, envelope.ts, envelope.payload);
  return crypto.verify(peerPublicKey, data, envelope.sig);
}

/** Serialize envelope to bytes. */
export function envelopeToBytes(envelope: BetweenEnvelope): Uint8Array {
  return new TextEncoder().encode(JSON.stringify(envelope));
}

/** Deserialize envelope from bytes. */
export function envelopeFromBytes(data: Uint8Array): BetweenEnvelope {
  return JSON.parse(new TextDecoder().decode(data)) as BetweenEnvelope;
}
